userspace/libscap/engine/nodriver/nodriver.c

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright (C) 2022 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "nodriver.h"
+
+#include "scap.h"
+#include "scap-int.h"
+#include "../common/strlcpy.h"
+#include "gettimeofday.h"
+#include "sleep.h"
+
+static struct nodriver_engine* alloc_handle(scap_t* main_handle, char* lasterr_ptr)
+{
+	struct nodriver_engine *engine = calloc(1, sizeof(struct nodriver_engine));
+	if(engine)
+	{
+		engine->m_lasterr = lasterr_ptr;
+	}
+	return engine;
+}
+
+static void free_handle(struct scap_engine_handle engine)
+{
+	free(engine.m_handle);
+}
+
+static int close_engine(struct scap_engine_handle engine)
+{
+	return SCAP_SUCCESS;
+}
+
+static int32_t next(struct scap_engine_handle handle, scap_evt** pevent, uint16_t* pcpuid)
+{
+	static scap_evt evt;
+	evt.len = 0;
+	evt.tid = -1;
+	evt.type = PPME_SCAPEVENT_X;
+	evt.nparams = 0;
+
+	sleep_ms(100);
+
+	evt.ts = get_timestamp_ns();
+	*pevent = &evt;
+	return SCAP_SUCCESS;
+}
+
+static int32_t start_capture(struct scap_engine_handle engine)
+{
+	return SCAP_SUCCESS;
+}
+
+static int32_t stop_capture(struct scap_engine_handle engine)
+{
+	return SCAP_SUCCESS;
+}
+
+static int32_t unimplemented_op(char* err, size_t err_size)
```

Comment:
Can we have a common definition for this? 
Perhaps even returning SCAP_UNIMPLEMENTED and let it be managed at top level.

Reply:
Good question. Is a central implementation of a single line function worth it? Probably. Though I guess we'd have to have a separate library to keep it (can't use libscap itself as it would create a circular dependency). It might make more sense once we have all the vtables and more dummy implementations for engines that don't support everything.

It could either be a null engine library (providing just the baseline no-op implementations of everything) or an engine_util collection (things like devset would go in there). Thoughts?

Also, I'm mildly worried that returning SCAP_UNIMPLEMENTED will cause a lot of failures when we expect SCAP_SUCCESS ~everywhere. I agree this is the right way but probably not in this PR set and IMHO certainly not in this PR.

Reply:
> It might make more sense once we have all the vtables and more dummy implementations for engines that don't support everything.

I agree!

> It could either be a null engine library (providing just the baseline no-op implementations of everything) or an engine_util collection (things like devset would go in there). Thoughts?

I'd go with the null engine library, i like the idea of doing something like (super pseudo code ofc):
```
if (I-can-manage-this) { manage(); return SCAP_SUCCESS; }
return null_engine.__func__();
```

Reply:
^^ This looks like it could live in the vtable dispatch code (if there's a vtable entry, call it, otherwise do the no-op default action), but then we pay the cost of checking the pointer for every call. Still, apart from next() (which could be special cased and just blindly call into the vtable) these should all be far from the hot path, so it's probably okay.

It wouldn't work for e.g. supporting just a subset of the settings but would save us from creating a separate tiny library. Thoughts?

Reply:
I think that having a noop library would allow engines to only implement a part of the symbols, using as fallback the noop (like:
```
const struct scap_vtable scap_nodriver_engine = {
	.name = "nodriver",
	.mode = SCAP_MODE_NODRIVER,

	...

	.start_capture = noop_engine_start_capture,
	.stop_capture = noop_engine_stop_capture,
	
        ...
};
```
Plus, it would still allow us to do the trick aforementioned when an unimplemented path is being processed.

Reply:
ok, done.